plugin/evm/warp_backend.go

```diff
@@ -0,0 +1,96 @@
+// (c) 2019-2020, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package evm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/utils/hashing"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/database/versiondb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/vms/platformvm/teleporter"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var (
+	_ WarpBackend = &WarpMessagesDB{}
+
+	dbPrefix = []byte("warp_messages")
+)
+
+const (
+	signatureCacheSize = 500
+)
+
+// WarpBackend keeps track of messages that are accepted by the warp precompiles and add them into a database.
+// The backend is also used to query for warp message signatures by the signature request handler.
+type WarpBackend interface {
+	// AddMessage is called in the precompile OnAccept, to add warp messages into the database.
+	AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error
+
+	// GetSignature returns the signature of the requested message hash.
+	GetSignature(ctx context.Context, messageHash ids.ID) ([]byte, error)
+}
+
+// WarpMessagesDB implements WarpBackend, keeping track of warp messages, and generating message signatures.
+type WarpMessagesDB struct {
+	database.Database
+	snowCtx        *snow.Context
+	signatureCache *lru.Cache
+}
+
+// NewWarpMessagesDB creates a new WarpMessagesDB, and initializes the signature cache and message tracking database.
+func NewWarpMessagesDB(snowCtx *snow.Context, vmDB *versiondb.Database) (*WarpMessagesDB, error) {
+	signatureCache, err := lru.New(signatureCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	db := &WarpMessagesDB{
+		Database:       prefixdb.New(dbPrefix, vmDB),
+		snowCtx:        snowCtx,
+		signatureCache: signatureCache,
+	}
+
+	return db, nil
+}
+
+func (w *WarpMessagesDB) AddMessage(ctx context.Context, unsignedMessage *teleporter.UnsignedMessage) error {
+	messageHashBytes := hashing.ComputeHash256(unsignedMessage.Bytes())
+	messageHash, err := ids.ToID(messageHashBytes)
+	if err != nil {
+		return fmt.Errorf("failed to add message with key %s to warp database: %w", messageHash.String(), err)
+	}
+
+	return w.Put(messageHash[:], unsignedMessage.Bytes())
```

Comment:
Would it make sense to sign the message here and place the signature in the database instead of keeping the whole message? We only need to fetch the signature so no need to store the full message.

Main question to me is the performance tradeoff.

Reply:
yup makes sense, made the change and added comment.

Reply:
hmm would this create a problem if node shutsdown then BLS key changes and boots again? I don't know how this will be used but this should not assume that BLS keys will never change on the node.

Reply:
hmm ye good question, I'm not sure if during the node's shutdown, whether the backend/cache would be wiped. So that after it comes back up with a different BLS key, it'd start from scratch.

Reply:
cache would be wiped bu db wont be. since we started to write signatures to db they will persist between restarts, and also in the case where BLS key changes. So any old signature will be available for the node even if the node would not be able to sign the bytes due to key change.

Reply:
Yeah this is definitely a valid concern. I think there are few considerations here with respect to storing the signature rather than the message (or message hash):
1. An honest node should never return an invalid signature. This could happen if we store signatures and then the BLS key changes. If we store signatures in the DB we'll need to be careful to invalidate them when we go to use them after changing a BLS key.
2. It's not the end of the world if an honest node is unable to return a signature for an older message after it's BLS key has changed. We only need 67% of nodes to sign for any given message.
3. Ideally, we'd avoid re-signing messages unnecessarily. 

Considering all of that, I think the cleanest option is to store a hash of each message, where the hash is what will be signed to generate the BLS signature of a message. We can still cache signatures that are generated since the cache will be cleared the restart if the BLS key changes. And we have the benefit of not needing to store arbitrary size messages to disk.

I think the simplicity of that approach makes it more desirable than storing the signature with some identifier of which key was used to generate the signature, and then checking that the key is still the current key when returning a signature from the DB.

Curious to hear other's thoughts though (cc @aaronbuchwald, @ceyonur, @minghinmatthewlam)

Reply:
I like the approach in terms of the most recently added messages will already have their signatures saved in the cache, and won't need to call on the db to resign the hash to generate a signature. Also, this would be simpler than having some identifier to check whether the saved signature in the db matches the current key of the node.

I agree with point 2, one node changing BLS key and stale db signatures we'd still only need 67%, but considering different sized subnets/nodes with big % stake in subnets, might be good to avoid this case.

Reply:
It makes much sense to use hashes for signatures rather than the full messages. In that case I think we would not need to store anything in the DB but just in the cache, right?